pkg/cluster/delete.go

```diff
@@ -405,30 +405,41 @@ func (m *manager) deleteFederatedCredentials(ctx context.Context) error {
 			return err
 		}
 
-		platformWIRole, exists := platformWIRolesByRoleName[name]
-		if !exists {
-			continue
+		federatedCredentials, err := m.clusterMsiFederatedIdentityCredentials.List(
+			ctx,
+			identityResourceId.ResourceGroup,
+			identityResourceId.ResourceName,
+			&armmsi.FederatedIdentityCredentialsClientListOptions{},
+		)
+		if err != nil {
+			return err
```

Comment:
Another potential bug, when a Cx during creation provides a platform identity with a missing role assignment on MSI to read the platform identity fails during populating the clientid/objectid.
But during deletion, it fails here too, this causes a blocker to delete.
Can we instead just console the error and allow the delete flow to continue??

> m.log.Errorf("failed to get federated identity credentials for %s: %v", identity.ResourceID, err.Error())
> continue

Reply:
Do you have specifics on what shape the error would take? I think the solution here would be to bail out and stop trying to delete any federated credentials, if we run into an authorization error when attempting to either list or delete credentials, but I want to make sure this manifests as an authorization error specifically before implementing that.

Reply:
Since you are bailing out here based on if OIDC populated or not, it won't be blocked for the example I provided above, but can happen in case the MSI's permission were revoked after cluster deletion or a new platform identity is added with MSI permission missing and then deletion is ran.

Below is an example of the error:

ERRO[2024-11-21T10:34:20-05:00]pkg/cluster/delete.go:415 cluster.(*manager).deleteFederatedCredentials() failed to get federated identity credentials for /subscriptions/<REDACTED>/resourceGroups/rajdchau-miwi/providers/Microsoft.ManagedIdentity/userAssignedIdentities/aro-NetworkOperator: GET https://management.azure.com/subscriptions/<REDACTED>/resourceGroups/rajdchau-miwi/providers/Microsoft.ManagedIdentity/userAssignedIdentities/aro-NetworkOperator/federatedIdentityCredentials
--------------------------------------------------------------------------------
RESPONSE 403: 403 Forbidden
ERROR CODE: AuthorizationFailed
--------------------------------------------------------------------------------
{
  "error": {
    "code": "AuthorizationFailed",
    "message": "The client '<REDACTED>' with object id '<REDACTED>' does not have authorization to perform action 'Microsoft.ManagedIdentity/userAssignedIdentities/federatedIdentityCredentials/read' over scope '/subscriptions/<REDACTED>/resourceGroups/rajdchau-miwi/providers/Microsoft.ManagedIdentity/userAssignedIdentities/aro-NetworkOperator' or the scope is invalid. If access was recently granted, please refresh your credentials."
  }
}
--------------------------------------------------------------------------------

Reply:
Thanks. I'll implement a non-failing short-circuit here if there is an auth error.

Reply:
or actually, I guess to stay in line with error handling during deletion, we should log and continue regardless of what the error is.

Reply:
Handled in latest commit.